lib/model/recent_senders.dart

```diff
@@ -0,0 +1,189 @@
+  /// index = somePrimes.possibleIndexOf(19); // -4
+  /// ```
+  int possibleIndexOf(T element) {
+    int low = 0;
+    int high = length - 1;
+
+    while (high >= low) {
+      int mid = (low + high) ~/ 2;
+      if (element == this[mid]) {
+        return mid;
+      } else if (element < this[mid]) {
+        high = mid - 1;
+      } else {
+        low = mid + 1;
+      }
+    }
+    return -low - 1;
+  }
+}
+
+/// A data structure to keep track of message ids.
+class IdTracker {
+  final List<int> _ids = [];
+
+  void add(int id) {
+    if (_ids.isEmpty) {
+      _ids.add(id);
+    } else {
+      int i = _ids.possibleIndexOf(id);
+      if (i >= 0) { // the [id] already exists, so do not add it.
+        return;
+      }
+      i = -i - 1; // change the index back to 0-based for readability.
+      if (i == _ids.length) { // the [id] should be added to the end of the list.
+        _ids.add(id);
+      } else {
+        _ids.insert(i, id);
+      }
+    }
+  }
+
+  // TODO: remove
+
+  /// The maximum id in the tracker list.
+  ///
+  /// Returns -1 if the tracker list is empty.
+  int get maxId => _ids.isNotEmpty ? _ids.last : -1;
+
+  @override
+  String toString() {
+    return _ids.toString();
+  }
+}
+
+/// A data structure to keep track of stream and topic messages.
+///
+/// The owner should call [clear] in order to free resources.
+class RecentSenders {
+  // streamSenders[streamId][senderId] = IdTracker
+  final Map<int, Map<int, IdTracker>> _streamSenders = {};
+
+  // topicSenders[streamId][topic][senderId] = IdTracker
+  final Map<int, Map<String, Map<int, IdTracker>>> _topicSenders = {};
+
+  void clear() {
+    _streamSenders.clear();
+    _topicSenders.clear();
+  }
+
+  int maxIdForStreamSender({required int streamId, required int senderId}) {
+    return _streamSenders[streamId]?[senderId]?.maxId ?? -1;
+  }
+
+  int maxIdForStreamTopicSender({
+    required int streamId,
+    required String topic,
+    required int senderId,
+  }) {
+    topic = topic.toLowerCase();
+    return _topicSenders[streamId]?[topic]?[senderId]?.maxId ?? -1;
+  }
+
+  void addStreamMessage({
+    required int streamId,
+    required int senderId,
+    required messageId,
+  }) {
+    final senderMap = _streamSenders[streamId] ?? {};
+    final idTracker = senderMap[senderId] ?? IdTracker();
+    _streamSenders[streamId] = senderMap;
+    senderMap[senderId] = idTracker;
+    idTracker.add(messageId);
+  }
+
+  void addTopicMessage({
+    required int streamId,
+    required String topic,
+    required int senderId,
+    required messageId,
+  }) {
+    topic = topic.toLowerCase();
+    final topicMap = _topicSenders[streamId] ?? {};
+    final senderMap = topicMap[topic] ?? {};
+    final idTracker = senderMap[senderId] ?? IdTracker();
+    _topicSenders[streamId] = topicMap;
+    topicMap[topic] = senderMap;
+    senderMap[senderId] = idTracker;
+    idTracker.add(messageId);
+  }
+
+  void processStreamMessage(Message message) {
```

Comment:
> Could fix by renaming to processMessage, or by narrowing the param's type to StreamMessage instead of Message.

Went with `processAsStreamMessage`. I thought when using `processMessage`, one would think that EACH message is processed; and when narrowing the param's type to `StreamMessage`, each caller should always make sure they're passing the correct message type by using a check, for example:
```dart
if (message is StreamMessage) {
  recentSenders.processStreamMessage(message);
}
```
I think it would be cleaner to add this check in the method itself and use a somewhat descriptive method name.
What do you think?

> Maybe we can also find a more specific verb than "process".

I think the verb "process" fits good in here as it extracts some data from the message and keeps track of it in a data structure. But I would appreciate a better verb from your side!